File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    PriorityQueue<Integer> maxHeap = new PriorityQueue<Integer>(Collections.reverseOrder());
    public int kthSmallest(TreeNode root, int k) {
        dfs(root, k);
        return maxHeap.peek();
    }

    private void dfs(TreeNode root, int k) {
        if (root == null) return;

        dfs(root.left, k);
        maxHeap.add(root.val);
        if (maxHeap.size() > k) {
            maxHeap.poll();
        }
        dfs(root.right, k);

        return;
    }
}

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        count = 0
        ans = 0
        def dfs(root):
            if not root:
                return
            
            dfs(root.left)
            nonlocal count
            count += 1
            if count == k:
                nonlocal ans
                ans = root.val
                return
            dfs(root.right)
            return

        dfs(root)
        return ans
        